File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:

        # 1, 5, 11, 5
        # sum = 22
        # total = sum // 2
        # algo - 
        # basically i need to find a subset whose sum is 11
        # if thats not possible, return false

        if sum(nums) % 2:
            return False

        dp = set()
        dp.add(0)
        target = sum(nums) // 2

        for i in range(len(nums)):
            nextDP = set()
            for t in dp:
                # if (t + nums[i]) == target:
                #     return True
                nextDP.add(t + nums[i])
                nextDP.add(t)
            dp = nextDP
        return True if target in dp else False

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {

        int sum = 0;
        for (int num : nums) {
            sum += num;
        }

        if (sum % 2 != 0) {
            return false;
        }

        int target = sum / 2;
        Set<Integer> dp = new HashSet<>();
        dp.add(0);

        for (int i = 0; i < nums.length; i++) {
            Set<Integer> nextDP = new HashSet<>(dp);

            if (nums[i] == target) {
                return true;
            }
            for (int j : dp ) {
                if (j + nums[i] == target) {
                    return true;
                }
                nextDP.add(nums[i]);
                nextDP.add(nums[i] + j);
            }
            dp = nextDP;
        }

        return false;
    }
}